crypto/provider.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/err.h>
+#include <openssl/cryptoerr.h>
+#include <openssl/provider.h>
+#include "internal/provider.h"
+
+OSSL_PROVIDER *OSSL_load_provider(OPENSSL_CTX *libctx, const char *name)
+{
+    OSSL_PROVIDER *prov = NULL;
+    OSSL_PROVIDER *created = NULL;
+
+    /* Find it or create it */
+    if ((prov = ossl_provider_find(libctx, name)) == NULL
+        && (prov = created = ossl_provider_new(libctx, name, NULL)) == NULL)
+        return NULL;
+
+    if (!ossl_provider_load(prov)) {
+        ossl_provider_free(created);
+        return NULL;
+    }
+
+    /*
+     * The provider object is safely in its store, no need for us
+     * to hold on to it
+     */
+
+    return prov;
+}
+
+int OSSL_unload_provider(OSSL_PROVIDER *prov)
+{
+    ossl_provider_free(prov);
+    return 1;
+}
+
+const OSSL_ITEM *OSSL_get_provider_param_types(OSSL_PROVIDER *prov)
+{
+    return ossl_provider_get_param_types(prov);
+}
+
+int OSSL_get_provider_params(OSSL_PROVIDER *prov, OSSL_PARAM params[])
+{
+    return ossl_provider_get_params(prov, params);
```

Comment:
Does the int return type to this mean that the whole operation succeeds or it fails? Lets say we want to query a provider for 10 different parameters, but the provider only knows about 4 of them. Is it expected to fail in this case? And does that mean we will be reduced to querying each one individually?

Reply:
The responder (I'm starting to use that term for a function that's supposed to respond to a param request) isn't expected to return an error for things it doesn't know.  It will simply ignore them.  The way for the caller to detect what was known to the responder is to see if `*return_size` has been modified.

Reply:
I wrote some recommendations in the `OSSL_PARAM` documentation, which I hope clarify this question a bit.  Please comment there if further clarification is needed.